### PR TITLE
[WIP] Internationalization Support

### PR DESCRIPTION
**Pull request by tsengia**
This PR will solve issue #297 by adding in internationalization support to Glowstone.
A `LanguageManager` class has been added to handle the retrieval of internationalized Strings.
It also provides a method to retrieve the internationalized String according to the server's settings.
The `default-locale` setting will set the language/locale to be used on the console and by default if `local-locale` is set to true. If `lock-locale` is set to false, the player's locale will be used to relay messages to them.
Currently there is only a few hard-coded messages that have been replaced with internationalized support, and more languages other than English needs to be added.

### CODE REVIEW

**Comment by Paulomart** [comment]
Good, but we need arguments that can be put into translated strings. Using the formatter would make sense. Eg: String::format

**Comment by tsengia** [comment]
@Paulomart noted, and will add in

**Comment by mastercoms** [comment]
Looks nice so far.

Since you're changing logger usage already, could you also make sure that Glowstone always uses logger.loglevel (like logger.info) for normal messages, and logger.log(Level, message, throwable) for error messages. Also, please search for usages of System.out.print and other logging functions if any and change them to use logger.

Also, sorry for the formatting messages (I hate dealing with formatting, too), but I have to wait until this PR is not WIP to comment on actual code.

Finally, it seems like you changed the permissions for the files that you modified (644 to 755). Please change them back if you can.

File: src/main/java/net/glowstone/entity/GlowPlayer.java [hunk 1]
```diff
@@ -295,7 +294,7 @@ public void playEffect(Location location, Effect effect, int id, int data, float
      * The one block the player is currently digging.
      */
     private GlowBlock digging;
-
+
```
Comment (mastercoms) [inline src/main/java/net/glowstone/entity/GlowPlayer.java:297]: Excess diff

File: src/main/java/net/glowstone/util/ServerConfig.java [hunk 2]
```diff
@@ -377,7 +379,7 @@ private boolean migrate() {
         PERMISSIONS_FILE("files.permissions", "permissions.yml", Migrate.BUKKIT, "settings.permissions-file"),
         COMMANDS_FILE("files.commands", "commands.yml"),
         HELP_FILE("files.help", "help.yml"),
-
+
```
Comment (mastercoms) [inline src/main/java/net/glowstone/util/ServerConfig.java:382]: Excess diff

File: src/main/java/net/glowstone/util/lang/LanguageManager.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    /*
+     * Returns the internationalized message according to the effective locale of the player.
+     */
+    public String getString(String key, GlowPlayer p) {
+        String locale = this.getEffectiveLocale(p);
+        ResourceBundle rb = ResourceBundle.getBundle(baseName, Locale.forLanguageTag(locale));
+        return rb.getString(key);
+    }
+    /**
```
Comment (mastercoms) [inline src/main/java/net/glowstone/util/lang/LanguageManager.java]: Add a new line

File: src/main/java/net/glowstone/util/lang/LanguageManager.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.ResourceBundle;
+
+import net.glowstone.GlowServer;
+import net.glowstone.entity.GlowPlayer;
+
+
+public class LanguageManager {
+    private String baseName = "lang/messages";
+    private ResourceBundle defaultBundle = ResourceBundle.getBundle(baseName, Locale.forLanguageTag(GlowServer.defaultLocale));
+    /*
```
Comment (mastercoms) [inline src/main/java/net/glowstone/util/lang/LanguageManager.java]: Add a new line

File: src/main/java/net/glowstone/GlowServer.java [hunk 5]
```diff
@@ -216,10 +219,24 @@
...
+    public static String defaultLocale;
+    /**
+     * Boolean if the default locale in the config should always be used.
+     */
+    public static boolean lockLocale;
+    /**
+     * The LanguageManager for the server. Used for internationalization.
+     */
+    public static LanguageManager languageManager;
     /**
```
Comment (mastercoms) [inline src/main/java/net/glowstone/GlowServer.java:237]: Add a new line

File: src/main/java/net/glowstone/GlowServer.java [hunk 6]
```diff
@@ -216,10 +219,24 @@
      * A view of all online players.
      */
     private final Set<GlowPlayer> onlineView = Collections.unmodifiableSet(onlinePlayers);
+    
+   /**
+    * The default locale to be used in the console and with the locked locale; read from the config.
+    */
+    public static String defaultLocale;
+    /**
```
Comment (mastercoms) [inline src/main/java/net/glowstone/GlowServer.java:228]: Add a new line

File: src/main/java/net/glowstone/GlowServer.java [hunk 7]
```diff
@@ -216,10 +219,24 @@
...
+     * Boolean if the default locale in the config should always be used.
+     */
+    public static boolean lockLocale;
+    /**
```
Comment (mastercoms) [inline src/main/java/net/glowstone/GlowServer.java:232]: Add a new line

**Comment by Paulomart** [comment]
Just an idea: On my server I have a Database with all my strings and the translations for them. It would be cool if plugins could do two things:
1. Hook into this to get and format messages etc.
2. Provide a source of translations from a database, a file etc. This would be checked first if null a fallback would be used.

**Comment by tsengia** [comment]
@Paulomart Or instead the user could install/edit their messages-yourLang.properties file to have the translated Strings for them (or download a translated messages-yourLang.properties file). Additionally, looking at the structure of the code, I think a plugin could easily extend the LanguageManager class and replace the default at start up, so only log messages before it would not be translated, but after it loads, everything would be translated by the plugin.

**Comment by Paulomart** [comment]
@ExpandingDev but theres no api for that. Even after you exteded the manager you need to set it. This would require to find all instances etc. Better to keep one instance and make it not extendable.

**Comment by tsengia** [comment]
@Paulomart The current LanguageManager is stored in GlowServer. There is only 1 GlowServer, and that is also statically typed, so the plugin could access the GlowServer, and use a setter to set the new, extended LanguageManager as the current LanguageManager. There is no need to keep track of every instance of LanguageManager classes. 
Here, some skeleton/protype code to explain:
https://gist.github.com/ExpandingDev/f0c5ce47835290e8d54967c183b44068

**Comment by tsengia** [comment]
@Paulomart Actually, the `lang` field is public, so it does not need a getter or setter, just a basic asignment with `=`

Or can the plugin not access the GlowServer? I thought it could, but if it cannot, then that's a problem.

**Comment by Paulomart** [comment]
If a plugin keeps it's own instance, like that: `LanguageManager lang = GlowServer.lang`.
And you set the GlowServer.lang field the plugin will still keep the old instance. This is not what we want. Just make an Interface LanguageProvider with getText(String key, String... values) and so on. The part that this is not in the API mhh asking @mastercoms

**Comment by tsengia** [comment]
@Paulomart ah, ok, didn't think of that. I will make the interface now

**Comment by Paulomart** [comment]
@ExpandingDev please see Mystalion/GlowstonePlusPlus@aec8f09f9e46e6126aef4d61cdb744cd4e08b1a3

**Comment by tsengia** [comment]
@Paulomart Oh, yeah that seems to be a better idea. Can we/I merge your commit into the PR somehow? Or should I merge your fork into my fork? Or make an entirely different PR for Glowstone?

**Comment by tsengia** [comment]
@Paulomart Nvm, got it

**Comment by Paulomart** [comment]
Ok then :)

**Pull request closed without merging**